### PR TITLE
Packagify autoscaler

### PR DESCRIPTION
**Pull request by saji-pivotal**
### What this PR does / why we need it

PR to convert autoscaler into a package and install it onto the management cluster when a workload cluster is being created

### Which issue(s) this PR fixes

Fixes #https://github.com/vmware-tanzu/tanzu-framework/issues/2320

### Describe testing done for PR
Manual Testing
1. Created and published the autoscaler package to my personal registry
2. Created a management-cluster on AWS
3. Set ENABLE_AUTOSCALER to true and created a workload cluster
4. Workload cluster creation was successful
5. Verified that the autoscaler deployment came up on the workload cluster
```
asudarshan-a01 ~/.config/tanzu/tkg/providers/ytt/03_customizations/autoscaler $ kubectl get pkgi -A
NAMESPACE    NAME                            PACKAGE NAME                               PACKAGE VERSION                            DESCRIPTION                                                            AGE
default      aws-cc-pi                       tkg-clusterclass-aws.tanzu.vmware.com      0.18.0-dev                                 Reconcile succeeded                                                    3d11h
default      test-wc-2-kapp-controller       kapp-controller.tanzu.vmware.com           0.34.0+vmware.1-tkg.1-zshippable           Delete failed: Error (see .status.usefulErrorMessage for details)      3h14m
default      test-wc-3-cluster-autoscaler    tkg-autoscaler.tanzu.vmware.com            0.21.0-dev                                 Reconcile succeeded                                                    2m15s
...


kubectl get deployments -A
NAMESPACE                           NAME                                            READY   UP-TO-DATE   AVAILABLE   AGE
capa-system                         capa-controller-manager                         1/1     1            1           3d11h
capi-kubeadm-bootstrap-system       capi-kubeadm-bootstrap-controller-manager       1/1     1            1           3d11h
capi-kubeadm-control-plane-system   capi-kubeadm-control-plane-controller-manager   1/1     1            1           3d11h
capi-system                         capi-controller-manager                         1/1     1            1           3d11h
cert-manager                        cert-manager                                    1/1     1            1           3d11h
cert-manager                        cert-manager-cainjector                         1/1     1            1           3d11h
cert-manager                        cert-manager-webhook                            1/1     1            1           3d11h
default                             test-wc-3-cluster-autoscaler                    1/1     1            1           2m28s
``` 

### Release note

```release-note
* Creates scaffolding for the autoscaler package
* Installs autoscaler on the management-cluster while a workload cluster is being created if the user enables it
```

### PR Checklist

- [ ] Squash the commits into one or a small number of logical commits
      
- [ ] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [ ] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

**Comment by vuil** [comment]
thanks for the changes, this PR needs a rebase. Please file an issue to describe what is being addressed and update the PR description with the testing done, fixes, and release notes as appropriate.

File: packages/management/tkg-autoscaler/bundle/config/upstream/tkg-autoscaler.yaml [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    matchLabels:
+      app: #@ "{}-cluster-autoscaler".format(data.values.CLUSTER_NAME)
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: #@ "{}-cluster-autoscaler".format(data.values.CLUSTER_NAME)
+    spec:
+      containers:
+        - image: tkg-autoscaler:latest
```
Comment (tenczar) [inline packages/management/tkg-autoscaler/bundle/config/upstream/tkg-autoscaler.yaml:46]: That can't be right. An unexpected pod restart shouldn't result in a change to the autoscale image.

File: pkg/v1/tkg/client/utils.go [hunk 2]
```diff
@@ -425,3 +427,70 @@ func getCCPlanFromLegacyPlan(plan string) (string, error) {
...
+
+func (c *TkgClient) GetAutoScalerValuesFileFromConfigs(clusterName, namespace, clusterRole, infrastructureProvider string) (string, error) {
+	autoScalerMap := map[string]string{
+		constants.ConfigVariableClusterName:  clusterName,
+		constants.ConfigVariableNamespace:    namespace,
+		constants.ConfigVariableClusterRole:  clusterRole,
+		constants.ConfigVariableProviderType: infrastructureProvider,
+	}
+
+	autoscalerMaxNodesTotal, err := c.TKGConfigReaderWriter().Get(constants.ConfigVariableAutoscalerMaxNodesTotal)
```
Comment (tenczar) [inline pkg/v1/tkg/client/utils.go:443]: we don't do any sanity checks on these values do we? someone could do `AUTOSCALER_MAX_NODES_TOTAL=banana` right? It might save us headaches later on to add some sanity checks.

**Review by tenczar — APPROVED** [review]
Up to you how much you want to dig into sanity checking the inputs. Definitely should make sure to add tests for those negative cases as well. I feel like I've just dumped an extra half day of work on you in the final hour...

File: packages/management/tkg-autoscaler/bundle/config/upstream/tkg-autoscaler.yaml (hunk 1 above)
Comment (saji-pivotal) [reply packages/management/tkg-autoscaler/bundle/config/upstream/tkg-autoscaler.yaml:46]: This is going to get replaced with a fixed version tag when we're building the package using `kbld`

File: packages/management/tkg-autoscaler/metadata.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+metadata:
+  name: tkg-autoscaler.tanzu.vmware.com
+  namespace: default
+spec:
+  displayName: "tkg-autoscaler"
+  longDescription: "Autoscaler package for tkg clusters"
+  shortDescription: "Defines the autoscaler resources for Tanzu"
+  providerName: VMware
+  maintainers:
+    - name: ""
```
Comment (anujc25) [inline packages/management/tkg-autoscaler/metadata.yaml:12]: Would be nice to add your name here.

File: packages/management/tkg-autoscaler/package.yaml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+apiVersion: data.packaging.carvel.dev/v1alpha1
+kind: Package
+metadata:
+  name: tkg-autoscaler.tanzu.vmware.com.0.21.0-dev
+  namespace: default
+spec:
+  refName: tkg-autoscaler.tanzu.vmware.com
+  version: 0.21.0-dev
```
Comment (anujc25) [inline packages/management/tkg-autoscaler/package.yaml:8]: Might be good to use `latest` as the version similar to other packages.

**Review by anujc25 — COMMENTED** [review]
How do we handle the upgrade of autoscaler during cluster upgrade? Currently I think we patch the deployment to use the updated autoscaler image. I am worried that if we are deploying this through packages, patching autoscaler deployment during upgrade might get overwritten by kapp controller reconciling packages. 

We need to test this upgrade scenario as well.

**Comment by saji-pivotal** [comment]
Closing this PR because it's pretty outdated. Will move these changes to a different PR
